### ISSUE TITLE
[Bug]: SQL Error "[HY000][1271] Illegal mix of collations for operation 'UNION'" after upgrade vom 10.6.X to 11.0.5

### ISSUE THREAD

**Issue by Fahl-Design** [pimcore/pimcore#15647]
### Pimcore version

11.0.5 (after upgrade)

### Steps to reproduce

- Have a working v10.6.X installation
- Upgrade to v11.0.5 
- Try to open an object
- See SQL error: `[HY000][1271] Illegal mix of collations for operation 'UNION'`
 
![image](https://github.com/pimcore/pimcore/assets/6690962/a364f893-1d9c-46cf-8036-1c3dad09af11)

trace: `\Pimcore\Model\DataObject\Objectbrick\Data\Dao::getRelationData` 

**cause: migration setting collate to "type" field in ONLY in objects table.** 
 https://github.com/pimcore/pimcore/blob/67b064e83c439b72d9d90bbff22ebc82cd56afc2/bundles/CoreBundle/src/Migrations/Version20221003115124.php#L34 

DDL
![image](https://github.com/pimcore/pimcore/assets/6690962/ed6b3513-34d4-4793-9a7b-ed598b1c02ea)

### Actual Behavior

See an SQL error:

```sql
SELECT r.dest_id as dest_id, r.dest_id as id, r.type, o.className as subtype, o.published as published, concat(o.path ,o.key) as `path` , r.index  as 'index'
FROM objects o, object_relations_Product r
WHERE r.fieldname="ElementDescriptionText"
  AND r.ownertype = 'object'
  AND r.src_id = 40146
  AND o.id = r.dest_id
  AND r.type='object'

UNION SELECT r.dest_id as dest_id, r.dest_id as id, r.type, a.type as subtype, 'NULL' as published, concat(a.path,a.filename) as `path`, r.index as 'index'
FROM assets a, object_relations_Product r
WHERE r.fieldname="ElementDescriptionText"
  AND r.ownertype = 'object'
  AND r.src_id = 40146
  AND a.id = r.dest_id
  AND r.type='asset'

UNION SELECT r.dest_id as dest_id, r.dest_id as id, r.type, d.type as subtype, d.published as published, concat(d.path,d.key) as `path`, r.index as 'index'
FROM documents d, object_relations_Product r
WHERE r.fieldname="ElementDescriptionText"
  AND r.ownertype = 'object'
  AND r.src_id = 40146
  AND d.id = r.dest_id
  AND r.type='document'

ORDER BY `index` ASC
```

Error:
`[HY000][1271] Illegal mix of collations for operation 'UNION'` (caused by "objects.type" field (set by migration Version20221003115124)

### Expected Behavior

no SQL error ;)

**Comment by robertSt7** [comment]
@Fahl-Design Could you please provide the exact Pimcore 10.6 version? Thanks

**Comment by Fahl-Design** [comment]
@robertSt7 10.6.4 to 11.0.5
(update 10.6.5 was released after I started the migration process(guide))

**Comment by cadeyrn** [comment]
> Could you please provide the exact Pimcore 10.6 version?

I have the same issue after upgrading Pimcore from version 10.5.20 to 10.6.6 and from there to 11.0.5. The objects of most classes can be opened, but the objects of two classes can not be opened due to this error.

**Comment by kingjia90** [comment]
not sure if the migration is the culprit, the error says about `Collation`, can you please check if all these 3 tables are `utf8mb4_general_ci` and if there's no custom collation for the `type` column?

**Comment by Fahl-Design** [comment]
@kingjia90 please check my screenshots and error description again, thx

*edit:
tl;dr:
"ONLY THE OBJECTS TABLE "type" has `utf8mb4_general_ci`

**Comment by kingjia90** [comment]
> @kingjia90 please check my screenshots and error description again, thx
> 
> *edit: tl;dr: "ONLY THE OBJECTS TABLE "type" has `utf8mb4_general_ci`

yes, that's helpful, but what are the tables' default collation? in my case they are all `utf8mb4_general_ci`, therefore it doesn't seem mattering that is defined as in `objects` or left empty.

**Comment by kingjia90** [comment]
> > Could you please provide the exact Pimcore 10.6 version?
> 
> I have the same issue after upgrading Pimcore from version 10.5.20 to 10.6.6 and from there to 11.0.5. The objects of most classes can be opened, but the objects of two classes can not be opened due to this error.

Mhh, then it's likely on the `object_relation_*` side.
Would a migration similar to https://github.com/pimcore/pimcore/blob/04e63bc7f7d434dc17f60ba3a61c1ef6f4d15862/bundles/CoreBundle/src/Migrations/Version20211103055110.php#L43 but with the `type` collation set to `utf8mb4_general_ci` work?

**Comment by Fahl-Design** [comment]
Well you have to options:
1) remove `utf8mb4_general_ci` form `objects.type` 
or 
2) add `utf8mb4_general_ci` to "assets" and "documents" and "object_realtion_*" type

![image](https://github.com/pimcore/pimcore/assets/6690962/eafb399b-98c7-4eb5-9a2a-c81c10aa2ddc)

they just have to match, and I do not really care which way :)

**Comment by kingjia90** [comment]
considering that if when the collation is not defined at column level (as `objects.type`), it would use the table collation. and the default is `utf8mb4_general_ci`
https://github.com/pimcore/pimcore/blob/194e370ca7529e3f0ff9ea55d51f023ed2f7e449/bundles/CoreBundle/Resources/config/pimcore/default.yaml#L91

Option 1 may not work because it's supposedly all already to `utf8mb4_general_ci`, even when not specified. like in the other tables beside `objects`. 
Migration down also doesn't give a hint on the collation change.

Can't tell unless you provide the current collation of all 4 tables.

**Comment by Fahl-Design** [comment]
ok, well if `utf8mb4_general_ci` is default why is it defined in some migrations like this one?

```
 $this->addSql("ALTER TABLE objects CHANGE o_type `type` enum('object','folder','variant') CHARACTER SET utf8mb4 COLLATE utf8mb4_general_ci DEFAULT NULL NULL;"); 
``` 

We started with Pimcore 5 or 6, and upgraded over the years to 10.6, and noch 11.X
Whatever was the default back than or was provided by one of your "premium" partners... 

anyway, 
you can assume they are different because of the SQL error this issue is about ;) 

since I am not alone with the *same* issue (cc @cadeyrn ) , and all needed information provided.
please go ahead and find a way to fix it or maybe ask @brusch if he can help you

thx
:v:

**Comment by kingjia90** [comment]
Cool, then we can go for option 1, by removing this explicit definition and keep to table default collation, then one has to solve the manually fix the tables collation difference. 
Thank you and wish you a nice weekend

**Comment by cadeyrn** [comment]
For the record: In my case the collation was set to utf8mb4_0900_ai_ci. After changing it to utf8mb4_general_ci in the type columns for the documents, assets and object_relations_* tables the error disappeared.

**Comment by dvesh3** [comment]
Fixed by #15696

**Comment by ghost** [comment]
Hi guys, 
I also had to modify the collation of objects.className to the table's default to fix this issue. Because in the union objects.className is used as subType and in my database the collations of objects.className, assets.type and documents.type where not the same. 

So after running these 2, this issue in my project was solved:

`ALTER TABLE objects CHANGE type `type` enum('object','folder','variant') DEFAULT NULL NULL;`
`ALTER TABLE objects CHANGE className className varchar(255) DEFAULT NULL NULL; `

**Comment by Hbitvof** [comment]
> Hi guys, I also had to modify the collation of objects.className to the table's default to fix this issue. Because in the union objects.className is used as subType and in my database the collations of objects.className, assets.type and documents.type where not the same.
> 
> So after running these 2, this issue in my project was solved:
> 
> `ALTER TABLE objects CHANGE type `type` enum('object','folder','variant') DEFAULT NULL NULL;` `ALTER TABLE objects CHANGE className className varchar(255) DEFAULT NULL NULL; `

I had the same problem after upgrading from 10.6 to 11.x. Running these two also solved our problem.
